sdk/servicebus/azure-servicebus/azure/servicebus/_common/auto_lock_renewer.py

```diff
@@ -91,23 +96,57 @@ def __init__(
         :type max_workers: Optional[int]
         """
         self._executor = executor or ThreadPoolExecutor(max_workers=max_workers)
+        # None indicates it's unknown whether the provided executor has max worker > 1
+        self._is_max_workers_greater_than_one = None if executor else (max_workers is None or max_workers > 1)
         self._shutdown = threading.Event()
-        self._sleep_time = 1
+        self._sleep_time = 0.5
         self._renew_period = 10
+        self._running = threading.Event()  # indicate whether the main worker thread is running
+        self._last_activity_timestamp = None  # the last timestamp when the main worker is active dealing with tasks
+        self._idle_timeout = 10  # the idle time that main worker thead should exist if there's no activity
         self._max_lock_renewal_duration = max_lock_renewal_duration
         self._on_lock_renew_failure = on_lock_renew_failure
+        self._renew_tasks = queue.Queue()  # type: ignore
+        self._infer_max_workers_time = 1
+        self._infer_max_workers_flags = [0, 0]
 
     def __enter__(self):
         if self._shutdown.is_set():
             raise ServiceBusError(
                 "The AutoLockRenewer has already been shutdown. Please create a new instance for"
                 " auto lock renewing."
             )
+
+        self._init_workers()
         return self
 
     def __exit__(self, *args):
         self.close()
 
+    def _init_workers(self):
+        self._infer_max_workers_greater_than_one_if_needed()
+        if not self._running.is_set():
+            self._running.set()
+            self._executor.submit(self._dispatch_worker)
+
+    def _infer_max_workers_greater_than_one_if_needed(self):
+        # infer max_workers value if executor is passed in
```

Comment:
would inferring max_workers by sleeping for 1 sec. (like below) increase the time it takes to run this? out of curiosity, is this insignificant enough of perf difference that we can ignore it?

also why do this rather than checking max workers with `executor._max_workers`? Is that bad by convention b/c we (as you mentioned in your description) only want to use public API?

Reply:
re1: it's a one time cost during the very beginning when auto lock renewer just starts, once the max_worker is inferred, `_infer_max_workers_greater_than_one_if_needed` would just be a no-op.

re2: right, the idea is not to rely on private ivars (hacky), which is unstable and could be changed by the python team.